schema/pgsql/schema.sql

```diff
@@ -21,6 +21,143 @@ CREATE TYPE comment_type AS ENUM ( 'comment', 'ack' );
+    ELSE
+      SELECT hard_state INTO last_hard_state
+      FROM service_state s
+      WHERE s.host_id = in_host_id
+        AND s.service_id = in_service_id;
+    END IF;
+  END IF;
+
+  IF last_hard_state IS NULL THEN
+    last_hard_state := 0;
+  END IF;
+
+  FOR cur IN
+    (
+      -- all downtime_start events before the end of the SLA interval
+      -- for downtimes that overlap the SLA interval in any way
+      SELECT
+        GREATEST(downtime_start, in_start_time) AS event_time,
+        'downtime_start' AS event_type,
+        1 AS event_prio,
+        NULL::tinyuint AS hard_state
+      FROM sla_history_downtime d
+      WHERE d.object_type = in_object_type
+        AND d.host_id = in_host_id
+        AND ((in_service_id IS NULL AND d.service_id IS NULL) OR d.service_id = in_service_id)
+        AND d.downtime_start < in_end_time
+        AND d.downtime_end >= in_start_time
+    ) UNION ALL (
+      -- all downtime_end events before the end of the SLA interval
+      -- for downtimes that overlap the SLA interval in any way
+      SELECT
+        downtime_end AS event_time,
+        'downtime_end' AS event_type,
+        2 AS event_prio,
+        NULL::tinyuint AS hard_state
+      FROM sla_history_downtime d
+      WHERE d.object_type = in_object_type
+        AND d.host_id = in_host_id
+        AND ((in_service_id IS NULL AND d.service_id IS NULL) OR d.service_id = in_service_id)
+        AND d.downtime_start < in_end_time
+        AND d.downtime_end >= in_start_time
+        AND d.downtime_end < in_end_time
+    ) UNION ALL (
+      -- all state events strictly in interval
+      SELECT
+        event_time,
+        'state_change' AS event_type,
+        0 AS event_prio,
+        hard_state
+      FROM sla_history_state s
+      WHERE s.object_type = in_object_type
+        AND s.host_id = in_host_id
+        AND ((in_service_id IS NULL AND s.service_id IS NULL) OR s.service_id = in_service_id)
+        AND s.event_time > in_start_time
+        AND s.event_time < in_end_time
+    ) UNION ALL (
+      -- end event to keep loop simple, values are not used
+      SELECT
+        in_end_time AS event_time,
+        'end' AS event_type,
+        3 AS event_prio,
+        NULL::tinyuint AS hard_state
+    )
+    ORDER BY event_time, event_prio
+  LOOP
+    IF ((in_object_type = 'host' AND last_hard_state > 0) OR (in_object_type = 'service' AND last_hard_state > 1))
+      AND last_hard_state != 99
+      AND active_downtimes = 0
+    THEN
+      problem_time := problem_time + cur.event_time - last_event_time;
+    END IF;
+
+    last_event_time := cur.event_time;
+    IF cur.event_type = 'state_change' THEN
+      last_hard_state := cur.hard_state;
+    ELSEIF cur.event_type = 'downtime_start' THEN
+      active_downtimes := active_downtimes + 1;
+    ELSEIF cur.event_type = 'downtime_end' THEN
+      active_downtimes := active_downtimes - 1;
+    END IF;
+  END LOOP;
+
+  total_time := in_end_time - in_start_time;
```

Comment:
What value would you expect if in your example, the host recovered in the middle of that 1 hour period (i.e. did not exist for almost a year, down for 30 minutes, up for 30 minutes)?

I see two approaches here:
1. Special-casing that if the state at the very beginning was UNKNOWN/99, use the first known state, so in above example, everything except 30 minutes would count as down, so value would be something like 0.x%.
2. Try to detect that the available data range is shorter then the requested range and truncate it accordingly, so 50% in the above example.

Number 2 would lead to different results between creating an object (thus seeing a UNKNOWN/99 state in the database so we know it was just created) and only enabling Icinga DB within the requested interval (data is just cut off and some value would be assumed for everything before).

Reply:
Data can only be truncated when removed from the stream, right? This means that Icinga DB is not running for a long time. I don't think we can solve that either.

Reply:
> I would go with `MAX(in_start_time, time_of_first_event)`.

Really for any event, not only if the previous state was UNKNOWN/99? This would then require having one event from before the report interval (or exactly at the beginning) so complicating retention/cleanup.

> However, we would still have ambiguities with decommissioned/recommissioned objects during the reporting period. But I don't think we can solve this without additional events.

I think excluding any time in a UNKNOWN/99 state should also produce acceptable results there. When something is recommissioned, it should start in that state again and the time it did not exist could be excluded from the total.

Reply:
> Really for any event, not only if the previous state was UNKNOWN/99? This would then require having one event from before the report interval (or exactly at the beginning) so complicating retention/cleanup.

We already select the initial state. I would also select the time of the initial state and use that for the `MAX(...`. There shouldn't be any downsides, right?

> I think excluding any time in a UNKNOWN/99 state should also produce acceptable results there. When something is recommissioned, it should start in that state again and the time it did not exist could be excluded from the total.

The problem is that you don't know when it was taken out of service, only when it was put back into service. So I wouldn't try to guess anything here as it can only be wrong.

Reply:
> Data can only be truncated when removed from the stream, right? This means that Icinga DB is not running for a long time. I don't think we can solve that either.

Data can be missing in the SQL database under the following circumstances:
1. At the beginning:
  a) You start using Icinga DB without migrating history from IDO.
  b) You are using the cleanup/retention feature.
  Both will work but there will be incorrect results for reports including earlier times.
2. At the end:
  a) Some of the latest history items is still processed at the moment. That should be no more than a few seconds though, so even if it's missing, it won't have a huge impact on the result.
  b) Icinga DB is currently not running. Then that interval become bigger, data is still cached in Redis though.
  All will be fine if things are running properly when generating the report.
3. Somewhere in the middle:
  This should only happen if you do things like stop Redis and then restart Icinga 2 while Redis is not running, so it will discard the pending history. That data is lost and there isn't anything we can do about this here.

Reply:
> We already select the initial state. I would also select the time of the initial state and use that for the `MAX(...`. There shouldn't be any downsides, right?

Then you can no longer just use `time < now-1y` to delete from the SLA history tables, otherwise when doing reports that start shortly after the beginning of the available data, you might reduce the total time for many hosts (it gets worse the more stable they are). I think that was the reason why we added `previous_hard_state` in the first place, so we can rely only on events within the report interval.